[docs] v1.14 Release Blogpost (#6098)
DevExpress/testcafe: docs/articles/blog/2021-3-31-testcafe-v1.14.0-released.md

---
layout: post
title: TestCafe v1.14.0 Released
permalink: /blog/:title.html
---
# TestCafe v1.14.0 Released

> This update includes new API methods for page scrolling, and a bugfix.

Reviewer: ```suggestion
This update includes a bugfix and new API methods for page scrolling.
```

<!--more-->

## Enhancements

### ⚙ Scroll Actions ([PR #6066](https://github.com/DevExpress/testcafe/pull/6066))

> When TestCafe interacts with elements on the page, it scrolls the page automatically to reach those elements.

Reviewer 2: What am I supposed to do with this information? Make it useful:

```suggestion
Users of earlier TestCafe versions had to interact with off-screen DOM elements to scroll the page.
```

> This release introduces actions that allow you to manually scroll the webpage and its elements.

Reviewer 2: all actions are manual actions

```suggestion
This release introduces dedicated scroll actions.
```

* [t.scroll](../documentation/reference/test-api/testcontroller/scroll.md) - scrolls the element to a specified position
* [t.scrollBy](../documentation/reference/test-api/testcontroller/scrollBy.md) - scrolls the element by the specified number of pixels
* [t.scrollIntoView](../documentation/reference/test-api/testcontroller/scrollintoview.md) - scrolls the element into view

> You can use the `t.scroll` action to scroll an element to a position:

Reviewer 2: These descriptions are inconsistent without a good reason.

```suggestion
The `t.scroll` action scrolls the target to reveal the specified coordinates:
```

```js
 import { Selector } from 'testcafe';

 fixture`Scroll Action`
     .page('http://example.com');

 test('Scroll the container', async t => {
     const container = Selector('#container');

     await t
         .scroll(container, 'bottomRight')
 });
 ```

>  `t.scrollBy` allows you to scroll an element (or the webpage) by a set amount of pixels. The example below scrolls the webpage 200px up and 500px to the right:

Reviewer 2: ```suggestion
 The `t.scrollBy` action scrolls the target by a set amount of pixels. The example below scrolls the webpage 200px up and 500px to the right:
```

 ```js
 fixture`Scroll Action`
     .page('http://example.com');

 test('Scroll the webpage', async t => {
     await t
         .scrollBy(500, -200)
 });
 ```

> Use `t.scrollIntoView` to scroll an element into view:

Reviewer 2: ```suggestion
The `t.scrollIntoView` action scrolls an element into view:
```

```js
 import { Selector } from 'testcafe';

 fixture `Scroll Actions`
     .page `http://www.example.com/`;

 test('Scroll element into view', async t => {
     const target = Selector('#target')

     await t
         .scrollIntoView(target)
 });
 ```

## Bug Fixes

> * Fixed an error that caused [expect.contains](../documentation/reference/test-api/testcontroller/expect/contains.md) assertions to display `undefined` for one of the diff values ([#5473](https://github.com/DevExpress/testcafe/issues/5473))

Reviewer: Is "diff" a tech term?

Author: ```suggestion
* Fixed an error that caused [expect.contains](../documentation/reference/test-api/testcontroller/expect/contains.md) assertions to display `undefined` for some the values in the comparison result ([#5473](https://github.com/DevExpress/testcafe/issues/5473))
```

> * Fixed an error that caused [expect.contains](../documentation/reference/test-api/testcontroller/expect/contains.md) assertions to display `undefined` for one of the passed string values in diffs ([#5473](https://github.com/DevExpress/testcafe/issues/5473))

Reviewer 2: simplified the phrasing a bit
```suggestion
* Fixed an error that caused [expect.contains](../documentation/reference/test-api/testcontroller/expect/contains.md) assertions to display `undefined` for one of the diff values ([#5473](https://github.com/DevExpress/testcafe/issues/5473))
```
